trainer/finetune.py

```diff
@@ -0,0 +1,240 @@
+
+  nltk_bs = nltk.translate.bleu_score.corpus_bleu(list_of_references=ref, hypotheses=hyp)
+  sb_bs = str(sacrebleu.corpus_bleu(predictions, [targets]))
+
+  print("NLTK BLEU SCORE: {:f}, SACREBLEU BLEU SCORE: {:s}, CHECKPOINT: {:d}".format(nltk_bs, sb_bs, checkpoint_step))
+  # Writes to $MODEL_DIR$/validation_eval/metrics$CHECKPOINT_NUMBER.json
+  metrics_path = os.path.join(
+          MODEL_DIR,
+          "validation_eval/metrics" + str(checkpoint_step) + ".json")
+  json.dump({"nltk_bleu_score" : nltk_bs, "sacrebleu_blue_score" : sb_bs, "recall@1" : 0}, tf.io.gfile.GFile(metrics_path, "w"))
+
+
+def main(argv):
+  """Main method for fintuning: builds, trains, and evaluates t5."""
+  tf.disable_v2_behavior()
+  tf.compat.v1.enable_eager_execution()
+
+  global PRETRAINED_DIR
+  global MODEL_DIR
+  PRETRAINED_DIR = os.path.join(BASE_PRETRAINED_DIR, FLAGS.size)
+  MODEL_DIR = os.path.join(MODELS_DIR, FLAGS.size, FLAGS.name)
+
+  print("--------DETECTING TPUs----")
+  try:
+    TPU_TOPOLOGY = "2x2"
+    tpu = tf.distribute.cluster_resolver.TPUClusterResolver()  # TPU detection
+    TPU_ADDRESS = tpu.get_master()
+    print('Running on TPU:', TPU_ADDRESS)
+  except ValueError:
+    raise BaseException('ERROR: Not connected to a TPU runtime')
+
+  # load or build data
+  if tf.io.gfile.exists(rd_counts_path):
+    print("TSV's Found")
+    # Used cached data and counts.
+    tf.logging.info("Loading Redial from cache.")
+    num_rd_examples = json.load(tf.io.gfile.GFile(rd_counts_path))
+  else:
+    print("TSV's Not Found")
+    # Create TSVs and get counts.
+    tf.logging.info("Generating Redial TSVs.")
+    num_rd_examples = {}
+    for split, fname in RD_SPLIT_FNAMES.items():
+      print(os.path.join(RD_JSONL_DIR, fname))
+      num_rd_examples[split] = preprocessing.rd_jsonl_to_tsv(
+          os.path.join(RD_JSONL_DIR, fname), rd_tsv_path[split])
+    json.dump(num_rd_examples, tf.io.gfile.GFile(rd_counts_path, "w"))
+
+  t5.data.TaskRegistry.add(
+      "rd_recommendations",
+      # Supply a function which returns a tf.data.Dataset.
+      dataset_fn=preprocessing.rd_dataset_fn,
+      splits=["train", "validation"],
+      # Supply a function which preprocesses text from the tf.data.Dataset.
+      text_preprocessor=[preprocessing.conversation_preprocessor],
+      # Use the same vocabulary that we used for pre-training.
+      sentencepiece_model_path=t5.data.DEFAULT_SPM_PATH,
+      # Lowercase targets before computing metrics.
+      postprocess_fn=t5.data.postprocessors.lower_text, 
+      # We'll use accuracy as our evaluation metric.
+      metric_fns=[t5.evaluation.metrics.accuracy, t5.evaluation.metrics.bleu],
+      # Not required, but helps for mixing and auto-caching.
+      num_input_examples=num_rd_examples)
+
+  # Set parallelism and batch size to fit on v2-8 TPU (if possible).
+  # Limit number of checkpoints to fit within 5GB (if possible).
+  model_parallelism, train_batch_size, keep_checkpoint_max = {
```

Comment:
Should these constants be coming from the T5 library or did you have to tune them yourself?

Reply:
These were from the notebook, I'm not sure if they from the t5 or not. I'll look into it.